File: app/src/main/java/com/example/dcmtkdemo/view/DicomUploadDialog.java
package com.example.dcmtkdemo.view;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.dcmtkdemo.R;
import com.example.dcmtkdemo.model.DicomImageRecord;
import com.example.dcmtkdemo.viewmodel.PacsViewModel;

import java.util.List;

public class DicomUploadDialog extends DialogFragment {

    private DicomUploadView uploadView;
    private PacsViewModel viewModel;
    private List<DicomImageRecord> uploadRecords;
    private OnUploadFinishedListener listener;
    private boolean autoStart = false;

    public interface OnUploadFinishedListener {
        void onFinished(int successCount, int totalCount);
    }

    public static DicomUploadDialog newInstance(List<DicomImageRecord> records) {
        return newInstance(records, false);
    }

    public static DicomUploadDialog newInstance(List<DicomImageRecord> records, boolean autoStart) {
        DicomUploadDialog dialog = new DicomUploadDialog();
        dialog.uploadRecords = records;
        dialog.autoStart = autoStart;
        return dialog;
    }

    public void setOnUploadFinishedListener(OnUploadFinishedListener listener) {
        this.listener = listener;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(false);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container
            , @Nullable Bundle savedInstanceState) {
        uploadView = new DicomUploadView(requireContext());
        uploadView.setBackgroundResource(R.drawable.popup_bg);
        return uploadView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(requireActivity()).get(PacsViewModel.class);

        uploadView.setConnectionInfo(
                viewModel.host.getValue(),
                viewModel.port.getValue(),
                viewModel.localAet.getValue(),
                viewModel.remoteAet.getValue()
        );
        uploadView.setUploadRecords(uploadRecords);
        uploadView.setOnUploadEventListener(new DicomUploadView.OnUploadEventListener() {
            @Override
            public void onFinished(int successCount, int totalCount) {
                if (listener != null) listener.onFinished(successCount, totalCount);
                dismiss();
            }

            @Override
            public void onCancel() {
                dismiss();
            }
        });

        if (autoStart) {
            uploadView.startUploadProcess();
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        Dialog dialog = getDialog();
        if (dialog != null) {
            dialog.setCancelable(false);
            dialog.setCanceledOnTouchOutside(false);
            Window window = dialog.getWindow();
            if (window != null) {
                int width = (int) (getResources().getDisplayMetrics().widthPixels * 0.90);
                window.setLayout(width, ViewGroup.LayoutParams.WRAP_CONTENT);
                window.setBackgroundDrawableResource(android.R.color.transparent);
            }
        }
    }
}

File: dcmtk/src/main/java/com/example/dcmtk/view/DicomUploadDialog.kt
package com.example.dcmtk.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.example.dcmtk.PacsManager
import com.example.dcmtk.model.DicomImageRecord
import com.example.dcmtk.model.PacsConfig
import com.example.dcmtk.viewmodel.PacsViewModel

/**
 * A dialog fragment that displays the DICOM upload progress.
 * It uses [DicomUploadView] to handle the actual upload process and UI.
 */
class DicomUploadDialog : DialogFragment() {

    private var uploadView: DicomUploadView? = null
    private lateinit var viewModel: PacsViewModel
    private var records: Array<DicomImageRecord>? = null
    private var listener: OnUploadFinishedListener? = null
    private var autoStart = false

    private var pacsConfig: PacsConfig? = null

    interface OnUploadFinishedListener {
        fun onFinished(successCount: Int, totalCount: Int)

        fun onItemStatus(index: Int, success: Boolean) {}
    }

    companion object {

        @JvmStatic
        fun newInstance(records: Array<DicomImageRecord>): DicomUploadDialog {
            return newInstance(records, false)
        }


        @JvmStatic
        fun newInstance(records: Array<DicomImageRecord>, autoStart: Boolean): DicomUploadDialog {
            return DicomUploadDialog().apply {
                this.records = records
                this.autoStart = autoStart
            }
        }

        @JvmStatic
        fun newInstance(
            records: Array<DicomImageRecord>,
            autoStart: Boolean,
            config: PacsConfig?
        ): DicomUploadDialog {
            return DicomUploadDialog().apply {
                this.records = records
                this.autoStart = autoStart
                this.pacsConfig = config
            }
        }


        @JvmStatic
        fun newInstance(
            records: Array<DicomImageRecord>,
            autoStart: Boolean,
            host: String?,
            port: Int,
            localAet: String?,
            remoteAet: String?
        ): DicomUploadDialog {
            return newInstance(records, autoStart, PacsConfig(host ?: ""
                , port, localAet ?: "", remoteAet ?: ""))
        }
    }

    fun setOnUploadFinishedListener(listener: OnUploadFinishedListener?) {
        this.listener = listener
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCancelable = false
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        viewModel = ViewModelProvider(requireActivity()).get(PacsViewModel::class.java)
        uploadView = DicomUploadView(requireContext(), pacsConfig ?: viewModel.pacsConfig.value
            , records).apply {
            // Background is already set in XML for the new view
        }
        return uploadView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        uploadView?.apply {
            setOnUploadEventListener(object : DicomUploadView.OnUploadEventListener {
                override fun onFinished(successCount: Int, totalCount: Int) {
                    listener?.onFinished(successCount, totalCount)
                }

                override fun onCancel() {
                    PacsManager.cancelOperation()
                    dismiss()
                }

                override fun onItemStatus(index: Int, success: Boolean) {
                    listener?.onItemStatus(index, success)
                }
            })

            if (autoStart) {
                startUploadProcess()
            }
        }
    }

    override fun onStart() {
        super.onStart()
        dialog?.apply {
            setCancelable(false)
            setCanceledOnTouchOutside(false)
            window?.apply {
                val width = (resources.displayMetrics.widthPixels * 0.45).toInt()
                setLayout(width, ViewGroup.LayoutParams.WRAP_CONTENT)
                setBackgroundDrawableResource(android.R.color.transparent)
            }
        }
    }
}
